File: sdk/typescript/src/blockchain/CrossChainBridge.ts
// Cross-chain bridge for TypeScript SDK
// Coordinates atomic operations between chat and currency chains

import axios, { AxiosInstance } from 'axios';
import { ChatChainClient } from './ChatChainClient';
import { CurrencyChainClient } from './CurrencyChainClient';

export enum CrossChainStatus {
  Pending = 'pending',
  ChatChainConfirmed = 'chat_chain_confirmed',
  CurrencyChainConfirmed = 'currency_chain_confirmed',
  AtomicSuccess = 'atomic_success',
  RolledBack = 'rolled_back',
  Failed = 'failed',
}

export interface CrossChainTransaction {
  id: string;
  operation: string;
  user_id: string;
  chat_chain_tx?: string;
  currency_chain_tx?: string;
  status: CrossChainStatus;
  created_at: number;
  finalized_at?: number;
}

export class CrossChainBridge {
  private client: AxiosInstance;

  constructor(
    _chatChain: ChatChainClient,
    _currencyChain: CurrencyChainClient,
    bridgeUrl: string = 'http://localhost:8548',
  ) {
    this.client = axios.create({
      baseURL: bridgeUrl,
      headers: { 'Content-Type': 'application/json' },
    });
  }

  async registerUserWithStake(
    userId: string,
    publicKey: Buffer,
    stakeAmount: number,
  ): Promise<CrossChainTransaction> {
    const response = await this.client.post('/register_user_with_stake', {
      user_id: userId,
      public_key: publicKey.toString('base64'),
      stake_amount: stakeAmount,
    });
    return response.data;
  }

  async createChannelWithFee(owner: string, channelName: string, creationFee: number): Promise<CrossChainTransaction> {
    const response = await this.client.post('/create_channel_with_fee', {
      owner,
      channel_name: channelName,
      creation_fee: creationFee,
    });
    return response.data;
  }

  async getStatus(bridgeTxId: string): Promise<CrossChainTransaction | null> {
    try {
      const response = await this.client.get(`/status/${bridgeTxId}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async getUserTransactions(userId: string): Promise<CrossChainTransaction[]> {
    const response = await this.client.get(`/user_transactions/${userId}`);
    return response.data;
  }

  async waitForAtomicCompletion(
    bridgeTxId: string,
    maxWaitMs: number = 60000,
  ): Promise<CrossChainTransaction> {
    const startTime = Date.now();
    while (Date.now() - startTime < maxWaitMs) {
      const tx = await this.getStatus(bridgeTxId);
      if (tx && tx.status === CrossChainStatus.AtomicSuccess) {
        return tx;
      }
      if (tx && (tx.status === CrossChainStatus.Failed || tx.status === CrossChainStatus.RolledBack)) {
        throw new Error(`Cross-chain transaction ${bridgeTxId} failed with status: ${tx.status}`);
      }
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    throw new Error(`Cross-chain transaction ${bridgeTxId} did not complete within ${maxWaitMs}ms`);
  }
}

export default CrossChainBridge;

File: sdk/dart/lib/src/blockchain/cross_chain_bridge.dart
/// Cross-chain bridge client for Dart SDK
/// Coordinates atomic operations between chat and currency chains

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'chat_chain_client.dart';
import 'currency_chain_client.dart';

/// Cross-chain transaction status
enum CrossChainStatus {
  pending,
  chatChainConfirmed,
  currencyChainConfirmed,
  atomicSuccess,
  rolledBack,
  failed,
}

/// Cross-chain transaction
class CrossChainTransaction {
  final String id;
  final String operation;
  final String userId;
  final String? chatChainTx;
  final String? currencyChainTx;
  final CrossChainStatus status;
  final int createdAt;
  final int? finalizedAt;

  CrossChainTransaction({
    required this.id,
    required this.operation,
    required this.userId,
    this.chatChainTx,
    this.currencyChainTx,
    required this.status,
    required this.createdAt,
    this.finalizedAt,
  });

  factory CrossChainTransaction.fromJson(Map<String, dynamic> json) {
    return CrossChainTransaction(
      id: json['id'] as String,
      operation: json['operation'] as String,
      userId: json['user_id'] as String,
      chatChainTx: json['chat_chain_tx'] as String?,
      currencyChainTx: json['currency_chain_tx'] as String?,
      status: CrossChainStatus.values[json['status'] as int],
      createdAt: json['created_at'] as int,
      finalizedAt: json['finalized_at'] as int?,
    );
  }
}

/// Bridge for coordinating atomic transactions between chains
class CrossChainBridge {
  final ChatChainClient chatChain;
  final CurrencyChainClient currencyChain;
  final String bridgeUrl;

  CrossChainBridge({
    required this.chatChain,
    required this.currencyChain,
    this.bridgeUrl = 'http://localhost:8548',
  });

  /// Register user with initial stake (atomic operation)
  Future<CrossChainTransaction> registerUserWithStake(
    String userId,
    List<int> publicKey,
    int stakeAmount,
  ) async {
    final response = await http.post(
      Uri.parse('$bridgeUrl/register_user_with_stake'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'user_id': userId,
        'public_key': base64Encode(publicKey),
        'stake_amount': stakeAmount,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to register user with stake: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return CrossChainTransaction.fromJson(data);
  }

  /// Create channel with fee (atomic operation)
  Future<CrossChainTransaction> createChannelWithFee(
    String owner,
    String channelName,
    int creationFee,
  ) async {
    final response = await http.post(
      Uri.parse('$bridgeUrl/create_channel_with_fee'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'owner': owner,
        'channel_name': channelName,
        'creation_fee': creationFee,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to create channel with fee: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return CrossChainTransaction.fromJson(data);
  }

  /// Get cross-chain transaction status
  Future<CrossChainTransaction?> getStatus(String bridgeTxId) async {
    final response = await http.get(
      Uri.parse('$bridgeUrl/status/$bridgeTxId'),
    );

    if (response.statusCode == 404) {
      return null;
    }

    if (response.statusCode != 200) {
      throw Exception('Failed to get status: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return CrossChainTransaction.fromJson(data);
  }

  /// Get all cross-chain transactions for user
  Future<List<CrossChainTransaction>> getUserTransactions(String userId) async {
    final response = await http.get(
      Uri.parse('$bridgeUrl/user_transactions/$userId'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to get transactions: ${response.body}');
    }

    final data = jsonDecode(response.body) as List<dynamic>;
    return data
        .map((item) => CrossChainTransaction.fromJson(item as Map<String, dynamic>))
        .toList();
  }
}
